### PR TITLE
♻️ Improve commutative unit reconstruction

### LINKED ISSUE #4052: Allow conversion of complex units to their standard SI representation
### Description of issue / requirement to address

Under our current unit system we heavily enforce strict SI with exceptions for some well know cases (power etc):
https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/develop/bluemira/base/constants.py#L268-L275

These only function for standalone units eg W not W/m^2.
It would be nice to do standard conversion to any combination of the above combined units

### Proposed solution

Modify these function:
https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/develop/bluemira/base/parameter_frame/_frame.py#L645-L685

to use the below access to the constituent units to reconstruct the unit as SI:

```python
import pint
u = pint.UnitRegistry()
q = 1 * u.watt/(u.meter**2)
for a in q.units._units.items():
     print(repr(u.Unit(a[0]).dimensionality), f'multiplier={a[1]}')
# <UnitsContainer({'[length]': 2, '[mass]': 1, '[time]': -3})> multiplier=1
# <UnitsContainer({'[length]': 1})> multiplier=-2
```

### Alternative solutions

Leave the current solution and we get:
```python
>>> q.to_base_units()
<Quantity(1.0, 'kilogram / second ** 3')>
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by je-cook**
## Linked Issues

Closes #4052 

## Description

This PR simplifies and refactors the unit reconstruction in parameterframe to more closely follow the original intent of the input units.

pint's `to_preferred` which works the same as our previous construction sometimes over collapses units as detailed in the issue this closes. I've worked around this by extracting our non commutative units and trying to reconstruct the remaining units into a more standard SI representation.

This needs to wait until pint has its next release, 0.25.3 (otherwise `to_preferred` can crash) and the #4154 as the newer versions of pint drop support for python3.10 (which that PR does).

To test you'll need to install pint from the repo. Probably a few more edge cases can be found to test too.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Review by CoronelBuendia — APPROVED** [review]
Just approving for the minor changes in geometry, equilibria, etc.

File: tests/base/parameterframe/test_parameterframe.py [hunk 1]
```diff
@@ -621,6 +634,15 @@ def test_wtf_units(self):
         assert frame.wtf3.value == pytest.approx(0.01388888)
         assert frame.wtf3.unit == "m⁵/deg/dpa/fpy/s"
 
+    def test_combined_units(self):
+        frame = CombinedFrame.from_dict(self.COMBINED_FRAME_DATA)
```
Comment (jonmaddock) [inline tests/base/parameterframe/test_parameterframe.py:648]: Is it worth saying that part of the purpose of this test is to check that this new unit reconstruction works? i.e. what the real purpose of this test is?

File: bluemira/base/parameter_frame/_units.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from pint import Quantity
+
+    from bluemira.base.parameter_frame._parameter import ParamDictT
+
+
+def _validate_units(param_data: ParamDictT, value_type: Iterable[type]):
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py]: I would ask for docstrings here, but hey, it's your code!

File: bluemira/base/parameter_frame/_units.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    try:
+        quantity = ureg.Quantity(param_data["value"], param_data["unit"])
+    except ValueError:
+        try:
+            quantity = ureg.Quantity(f"{param_data['value']}*{param_data['unit']}")
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:53]: Does this actually ever make a difference over the previous try statement?

File: bluemira/base/parameter_frame/_units.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        except pint.errors.PintError as pe:
+            if param_data["value"] is None:
+                quantity = ureg.Quantity(
+                    1 if param_data["unit"] in {None, ""} else param_data["unit"]
+                )
+            else:
+                raise ValueError("Unit conversion failed") from pe
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:61]: Could be worth commenting what the various try/except cases are and when they might be hit: some obvious, some not.

File: bluemira/base/parameter_frame/_units.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        else:
+            param_data["value"] = quantity.magnitude
+        param_data["unit"] = quantity.units
+    except KeyError as ke:
+        raise ValueError("Parameters need a value and a unit") from ke
+    except TypeError:
+        if param_data["value"] is None:
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:69]: This looks similar to above: when would this be hit instead of `PintError`?

File: bluemira/base/parameter_frame/_units.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            # dummy for None values
+            quantity = ureg.Quantity(
+                1 if param_data["unit"] in {None, ""} else param_data["unit"]
+            )
+        elif isinstance(param_data["value"], bool | str):
+            param_data["unit"] = "dimensionless"
+            return param_data
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:76]: Why return here, instead of running `_ensure_SI_unit_system()`?

File: bluemira/base/parameter_frame/_units.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        else:
+            raise
+
+    return _ensure_SI_unit_system(quantity, param_data, value_type)
+
+
+def _ensure_SI_unit_system(
+    quantity: Quantity, param_data: ParamDictT, value_type: Iterable[type]
+) -> ParamDictT:
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:85]: Docstring please.

File: bluemira/base/parameter_frame/_units.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    if quantity.units != ureg.dimensionless:
+        unit_q = _remake_units(quantity)
+    else:
+        unit_q = quantity
+
+    if unit_q.units != quantity.units and param_data["value"] is not None:
+        val = raw_uc(quantity.magnitude, quantity.units, unit_q.units)
+        if isinstance(param_data["value"], int) and int in value_type:
+            val = int(val)
+        param_data["value"] = val
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:110]: Don't understand what this is doing.

File: bluemira/base/parameter_frame/_units.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+    param_data["unit"] = f"{unit_q.units:~P}"
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:112]: ?

File: bluemira/base/parameter_frame/_units.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    The quantity of the conversion is not important here.
+    We just want the custom SI version of the input
+    The value conversion is done later.
+    """
+    unit_list, ind_list, fix_list = [], [], []
+    for no, (unit, multiplier) in enumerate(quantity.units._units.items()):
+        q = ureg.Quantity(unit)
+        if q.units not in {ureg.fpy, ureg.dpa}:
+            q = q.to_preferred()  # individually preferred (SI)
+        if q.units in {ureg.fpy, ureg.dpa} or not q.dimensionality:
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:141]: Combine with previous if statement?

File: bluemira/base/parameter_frame/_units.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            # is not commutative or is angle (as their dimensionless in pint)
+            fix_list.append(no)
+        else:
+            ind_list.append(no)
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:146]: What case is this for?

File: bluemira/base/parameter_frame/_units.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+    filtered_list = [unit_list[i] for i in filter_index]
+    for no, i in enumerate(filtered_list):
+        if not i.dimensionality and get_key(i) != ureg.dpa:
+            if get_key(i) in {ureg.steradian, ureg.square_degree}:
+                raise NotImplementedError("Solid angle conversion not supported")
+
+            filtered_list[no] = i.to(ureg.Unit(f"{ANGLE}**{get_exp(i)}"))
+
+    return math.prod(filtered_list)
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:180]: ?

File: bluemira/base/parameter_frame/_units.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+    return _combine_commutative(unit_list, ind_list) * _convert_non_commutative(
+        unit_list, fix_list
+    )
+
+
+def _convert_non_commutative(
+    unit_list: list[Quantity], filter_index: list[int]
+) -> Quantity:
+    """Converts angle units and combines non commutative units"""  # noqa: DOC201
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py]: Better definition of commutative would help here I think, especially an example.

File: bluemira/base/parameter_frame/_units.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+    Notes
+    -----
+    Uses pints milp optimisation which can over combine eg W/m^2 -> kg/s^3.
+    The optimisation minimises combinations eg the "knapsack problem".
+    This function prioritise shorter original over new which could be improved in future
+    """  # noqa: DOC201
+    filtered_list = [unit_list[i] for i in filter_index]
+
+    quantity = math.prod(filtered_list)
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:197]: Clearly I'm missing something...

File: bluemira/base/parameter_frame/_units.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    if not isinstance(quantity, ureg.Quantity):
+        return quantity
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:200]: When would this be the case? Hasn't this been checked already?

File: bluemira/base/parameter_frame/_units.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+    pq = quantity.to_preferred()
+    if len(quantity._units.keys()) <= len(pq._units.keys()):
+        return quantity
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:207]: Wouldn't we always want preferred units?

File: bluemira/base/parameter_frame/_units.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    The optimisation minimises combinations eg the "knapsack problem".
+    This function prioritise shorter original over new which could be improved in future
+    """  # noqa: DOC201
+    filtered_list = [unit_list[i] for i in filter_index]
+
+    quantity = math.prod(filtered_list)
+    if not isinstance(quantity, ureg.Quantity):
+        return quantity
+
+    pq = quantity.to_preferred()
```
Comment (jonmaddock) [inline bluemira/base/parameter_frame/_units.py:205]: Maybe highlight this: this one line does almost all of the work of unit conversion using pint: the rest is handling the stuff pint doesn't handle, right?

**Review by jonmaddock — CHANGES_REQUESTED** [review]
No problem with code, just questions, some of the answers to which might be worth turning into comments. This is quite tricky to get right (as you've done) and to understand working around pint's limitations for edge cases: some good examples as comments might help remove ambiguity and work out what's being done at each step.

File: bluemira/base/parameter_frame/_units.py (hunk 3 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:53]: yes the first fails whereas the second works:
```python
from pint import Quantity

try:
    Quantity(1, '1e-19m^2')
except ValueError as v:
    print(v)

Quantity('1*1e-19m^2')
```

File: bluemira/base/parameter_frame/_units.py (hunk 5 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:69]: see additional comments

File: bluemira/base/parameter_frame/_units.py (hunk 6 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:76]: It shortcircuits a load of logic that will never be used. Whereas in one case where we collapse a unit to dimensionless the quantity may now be different and therefore modify the value in some instances

File: bluemira/base/parameter_frame/_units.py (hunk 8 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:110]: I've added a comment

File: bluemira/base/parameter_frame/_units.py (hunk 9 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:112]: this is a short string-ified representation of the unit, removes pint objects from the user

File: bluemira/base/parameter_frame/_units.py (hunk 10 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:141]: q has possibly been updated in the first if statement and we need to deal with also this needs running if q has angles in it irrelevant of the previous step

File: bluemira/base/parameter_frame/_units.py (hunk 11 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:146]: commutative parts of the unit which we just want to pass through

File: bluemira/base/parameter_frame/_units.py (hunk 12 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:180]: multiplies all quantities together in the list

File: bluemira/base/parameter_frame/_units.py (hunk 14 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:197]: see previous comment on the same code

File: bluemira/base/parameter_frame/_units.py (hunk 15 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:200]: this covers if the multiplication became a raw number

File: bluemira/base/parameter_frame/_units.py (hunk 16 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py:207]: preffered units sometimes overcollapse the unit and in some cases the original unit was "better" eg W/m² can resolve to kg/s3

File: tests/base/parameterframe/test_parameterframe.py (hunk 1 above)
Comment (je-cook) [reply tests/base/parameterframe/test_parameterframe.py:648]: docstring updated

File: bluemira/base/parameter_frame/_units.py (hunk 13 above)
Comment (je-cook) [reply bluemira/base/parameter_frame/_units.py]: I've added some notes here that should help the definition

**Comment by josieapeters** [comment]
LGTM! tests all work well for me

**Review by josieapeters — APPROVED** [review]

**Review by jonmaddock — APPROVED** [review]
I'm happy with these comments: they answer all my queries.
